pkg/utils/util.go

```diff
@@ -9,3 +15,63 @@ func Contains(slice []string, value string) bool {
 	}
 	return false
 }
+
+// GetToolchainClusterName produces a name for ToolchainCluster object that is both deterministic and "reasonably unique".
+// The `ordinal` must be greater than 0 if there are multiple ToolchainCluster objects pointing to the same cluster. This
+// needs to be determined by the caller prior to calling this method.
+func GetToolchainClusterName(clusterType, serverAPIEndpoint, suffix string) (string, error) {
+	// NOTE: this function is ported from the original add-cluster.sh script to produce the same names during the transition
+	// period to the new operator-based approach to the member registration.
+	// Since add-cluster.sh was a bash script with a long history, the logic is a bit convoluted at places (especially in
+	// handling the numerical suffix).
+
+	// we need to make sure that:
+	// 1) the name is at most 63 characters long
+	// 2) the variable part is (a part of) the cluster hostname
+	// 3) it ends with a digit (supplied by the ordinal param) if it was shortened
+
+	suffix = strings.TrimSpace(suffix)
+
+	// the name always contains the cluster type, a hypen between the cluster type and the numerical suffix (if needed)
+	// Interestingly, this is computed BEFORE we determine if we need the numerical suffix at all, but that's the logic
+	// in the original script.
+	fixedLength := len(clusterType) + len(suffix) + 1
+
+	maxAllowedClusterHostNameLen := 62 - fixedLength // I think 62 is here, because we might default the ordinal to 1 later on
+
+	clusterHostName, err := sanitizeEnpointForUsageAsName(serverAPIEndpoint)
+	if err != nil {
+		return "", fmt.Errorf("failed to sanitize the endpoint for naming purposes: %w", err)
+	}
+	if len(clusterHostName) >= maxAllowedClusterHostNameLen {
+		clusterHostName = clusterHostName[0:maxAllowedClusterHostNameLen]
+		// the original script uses this approach to ensure that the name ends with an alphanumeric
+		// character (i.e. that the name doesn't end with a '.' after shortening the hostname)
+		if len(suffix) == 0 {
+			suffix = "1"
+		}
+	}
+
+	if len(suffix) == 0 {
+		return fmt.Sprintf("%s-%s", clusterType, clusterHostName), nil
+	}
+	return fmt.Sprintf("%s-%s%s", clusterType, clusterHostName, suffix), nil
```

Comment:
This would change the behavior. The original script really only re-initialized the empty suffix if the length was too long (which makes sense if we want to make sure that the name doesn't end with a '.' after chopping off the excessive chars from the hostname).

I really would not like to change the semantics of this because we need to be able to apply changes to already registered clusters using this command (see the checks for the matching object names in the validation) and changing the way the names are created could stop us from being able to do that.

Reply:
Ok I see, makes sense! Let's keep it as is 👍. Thanks!